File: backend/src/main/java/com/saif/portfolio/model/Blog.java
package com.saif.portfolio.model;

import java.time.Instant;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "blogs")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Blog {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @Column(nullable = false)
    private String title;

    @Column(nullable = false, unique = true)
    private String slug;

    @Column(nullable = false, length = 1000)
    private String summary;

    @Column(nullable = false, length = 100)
    private String category;

    @Column(name = "read_time", nullable = false, length = 50)
    private String readTime;

    @Column(nullable = false)
    private String author;

    @Column(name = "created_at", updatable = false)
    private Instant createdAt;

    @Column(name = "updated_at")
    private Instant updatedAt;

    @Column(columnDefinition = "TEXT")
    private String content;

    @OneToOne(mappedBy = "blog", cascade = CascadeType.ALL, orphanRemoval = true,fetch = FetchType.LAZY)
    private BlogImage image;

    // ✅ Add this part
    @PrePersist
    public void onCreate() {
        this.createdAt = Instant.now();
        this.updatedAt = Instant.now();
    }

    @PreUpdate
    public void onUpdate() {
        this.updatedAt = Instant.now();
    }
}

File: frontend/src/types/blog.ts
import { ImageUploadResponse } from "./project";

// Blog response from API
export interface BlogResponse {
  id: number;
  title: string;
  slug: string;
  summary: string;
  content: string;
  category: string;
  readTime: string;
  author: string;
  createdAt: string;
  updatedAt: string;
  image: ImageUploadResponse;
}

// Request payload for creating/updating blog
export interface BlogRequest {
  title: string;
  slug: string;
  summary: string;
  content: string;
  category: string;
  readTime: string;
  author: string;
  image: ImageUploadResponse;
}

// State types for blog pages
export interface BlogFilters {
  category?: string;
}